network/iptables.go

```diff
@@ -16,145 +16,273 @@
 	}
 }
 
-func ensureIPTables(ipt IPTables, rules []IPTablesRule) error {
-	exists, err := ipTablesRulesExist(ipt, rules)
-	if err != nil {
-		return fmt.Errorf("Error checking rule existence: %v", err)
-	}
-	if exists {
-		// if all the rules already exist, no need to do anything
-		return nil
+func (c Config) createRules(ipt IPTables) error {
+	for _, chain := range chains {
+		err := ipt.ClearChain(chain.table, chain.name)
+		if err != nil {
+			glog.Info("Clear chain ", chain, " failed: ", err)
+		}
 	}
-	// Otherwise, teardown all the rules and set them up again
-	// We do this because the order of the rules is important
-	log.Info("Some iptables rules are missing; deleting and recreating rules")
-	teardownIPTables(ipt, rules)
-	if err = setupIPTables(ipt, rules); err != nil {
-		return fmt.Errorf("Error setting up rules: %v", err)
+
+	for _, rule := range c.generateRules(ipt, c.Masquerade) {
+		glog.Info("Adding iptables rule: table: ", rule.table, " chain: ", rule.chain, " spec: ",
+			strings.Join(rule.getRule(), " "))
+
+		err := ipt.AppendUnique(rule.table, rule.chain, rule.getRule()...)
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func setupIPTables(ipt IPTables, rules []IPTablesRule) error {
-	for _, rule := range rules {
-		log.Info("Adding iptables rule: ", strings.Join(rule.rulespec, " "))
-		err := ipt.AppendUnique(rule.table, rule.chain, rule.rulespec...)
+func (c Config) rulesOk(ipt IPTables) error {
+	for _, rule := range c.generateRules(ipt, c.Masquerade) {
+		exists, err := ipt.Exists(rule.table, rule.chain, rule.getRule()...)
 		if err != nil {
-			return fmt.Errorf("failed to insert IPTables rule: %v", err)
+			return trace.Wrap(err)
+		}
+		if !exists {
+			return trace.NotFound("missing rule")
 		}
 	}
-
 	return nil
 }
 
-func teardownIPTables(ipt IPTables, rules []IPTablesRule) {
-	for _, rule := range rules {
-		log.Info("Deleting iptables rule: ", strings.Join(rule.rulespec, " "))
-		// We ignore errors here because if there's an error it's almost certainly because the rule
-		// doesn't exist, which is fine (we don't need to delete rules that don't exist)
-		ipt.Delete(rule.table, rule.chain, rule.rulespec...)
+func (c Config) ensureRules(ipt IPTables) {
+	err := c.rulesOk(ipt)
+	if err != nil && !trace.IsNotFound(err) {
+		glog.Error("Error checking iptables rules: ", trace.DebugReport(err))
+		return
 	}
+
+	if trace.IsNotFound(err) {
+		glog.Error("Discovered missing iptables rules, re-creating...")
+		// rules appear to be missing
+		// so we delete then recreate our rules
+		c.cleanupRules(ipt)
+
+		err = c.createRules(ipt)
+		if err != nil {
+			glog.Error("Error creating iptables rules: ", trace.DebugReport(err))
+		}
+	}
+}
+
+func (c Config) SetupAndEnsureIPTables() error {
+	ipt, err := iptables.New()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+
+		c.cleanupRules(ipt)
+		_ = c.createRules(ipt)
```

Comment:
Ignored error?

Reply:
Yea, I'm on the fence, this is really a peremptory creation so the rules get created immediately, but then the reconciliation loop would detect and report any discrepancies.

Reply:
Let's log it at least.